### ISSUE TITLE
Share workflow

### ISSUE THREAD

**Issue by nikkolasg** [drand/drand#343]
From @hsanjuan in [PR](https://github.com/drand/drand/pull/252#discussion_r426426111):
>The current drand share is a brittle process now:
    - Must ensure full connectivity outgoing and incoming
    - Must make sure generate-keypair has been run with the correct hostname
    - Must ensure everyone does it at the same time
    - The --leader part of the command blocks but it can also be cancelled and the sharing process continues in the background, but then it can be run again and I have no idea what happens (does it restart? Does it continue? Do I need to restart the drand node?)

>    I have seen it fail with really weird errors and then work after a couple of retries and restart of the nodes.
    It is very difficult to build anything automatically deployable with this
    The whole process need to be done in sync by everyone. It if fails, the whole "let's relaunch this thing needs to be done in sync by everyone.
To have a generator which takes a number of public keys and creates the group.toml file seems a much better approach. The configuration can be deployed following standard methods, not everyone needs to do it at the same time, connectivity problems can be debugged separately without affecting the process for everyone else, there is no need to elect a leader and have a special role...

**Comment by nikkolasg** [comment]
Some answers to your points.
> Must ensure full connectivity outgoing and incoming

That is anyway needed for the DKG - it's safe to assume that for the step before to create the DKG group.

> - Must make sure generate-keypair has been run with the correct hostname

That's another issue of having the endpoint being tied to the public key - and if we can remove that decoupling smoothly then  i'm all for it.

> -  - Must ensure everyone does it at the same time

That's not correct - The coordinator needs to be the first but after that, there is no need to sync every other operators together, they run the process when they can.

> The --leader part of the command blocks but it can also be cancelled and the sharing process continues in the background, but then it can be run again and I have no idea what happens (does it restart? Does it continue? Do I need to restart the drand node?)

Agreed, the cancellation is to be improved - the way it works is that the CLI commands sends a gRPC call to the daemon and the daemon call waits for the setup phase to finish before returning the gorup to the CLI endpoint. If the CLI endpoint is cancelled, the daemon will send it to the connection but it will fail, but the DKG still happenned.

> I have seen it fail with really weird errors and then work after a couple of retries and restart of the nodes.

If you have concrete scenario to show me where it fail, it would help !

> It is very difficult to build anything automatically deployable with this

Again, drand is **not** meant to be deployed in a cluster, not at all, so having a workflow easy for human operator is more important than having a workflow easy for machines.

> The whole process need to be done in sync by everyone. It if fails, the whole "let's relaunch this thing needs to be done in sync by everyone.

No, not in sync. And yes, if it fails, we have to relaunch this again. Maybe there could be some way to automate a retry...

> To have a generator which takes a number of public keys and creates the group.toml file seems a much better approach.

I tried that and from my experience with human operators, I strongly disagree (and i believe the other operators too from the feedback I've received.

> The configuration can be deployed following standard methods, 

What standard methods ? 

> not everyone needs to do it at the same time, 

As now.

> connectivity problems can be debugged separately without affecting the process for everyone else,

Each operator should check its connectivity before running the share command with the `drand util check <address>` CLI. 

> there is no need to elect a leader and have a special role...

There still needs to have a leader that starts the DKG, and before, when we had to manually download the group file, it was even worse because the leader had to wait until everyone did run the share command , so there was a much higher need to sync between operators.

**Comment by hsanjuan** [comment]
> There still needs to have a leader that starts the DKG, and before, when we had to manually download the group file, it was even worse because the leader had to wait until everyone did run the share command

Ok I missed this part. Is there a way to write a program that: given N public keys ouputs the group.toml file in an offline mode?

In a way that no one has to run the share command?




> > * Must ensure everyone does it at the same time
> 
> That's not correct - The coordinator needs to be the first but after that, there is no need to sync every other operators together, they run the process when they can.

As long as the operator has not restarted the process, or it has died itself for some reason.



> > The --leader part of the command blocks but it can also be cancelled and the sharing process continues in the background, but then it can be run again and I have no idea what happens (does it restart? Does it continue? Do I need to restart the drand node?)
> 
> Agreed, the cancellation is to be improved - the way it works is that the CLI commands sends a gRPC call to the daemon and the daemon call waits for the setup phase to finish before returning the gorup to the CLI endpoint. If the CLI endpoint is cancelled, the daemon will send it to the connection but it will fail, but the DKG still happenned.

So the program stays in setup mode forever if the share-phase is never completed?



> > I have seen it fail with really weird errors and then work after a couple of retries and restart of the nodes.
> 
> If you have concrete scenario to show me where it fail, it would help !

I think it happened like this: Given 2 nodes A, B. A is leader and starts the sharing. B contacts A fine (`drand util check` works). But A cannot contact B due to misconfiguration. On B: `drand share`, hangs or fails. Fix misconfiguration. On B: `drand share` does not work. However A is still not finished. Restart everything and then things work.


> > It is very difficult to build anything automatically deployable with this
> 
> Again, drand is **not** meant to be deployed in a cluster, not at all, so having a workflow easy for human operator is more important than having a workflow easy for machines.

I understand this, but anything that requires running CLI commands interactively is not fit for automatic deployment. It seems not possible to know what configuration the drand nodes need to use without running the drand nodes interactively. This is suboptimal because it requires access from the operator to the machines which are supposed to be super safe to run commands.



> > The whole process need to be done in sync by everyone. It if fails, the whole "let's relaunch this thing needs to be done in sync by everyone.
> 
> No, not in sync. And yes, if it fails, we have to relaunch this again. Maybe there could be some way to automate a retry...

Yes, in sync because a node needs to start DKG before anyone else, and that node needs to survive through the process and no node can be correctly configured without having to be online at the same time as the other nodes.

> > The configuration can be deployed following standard methods,
> 
> What standard methods ?

Anything that configures a node from scratch without having an interactive phase in which all operators need to perform actions and keep their nodes running. I am just aiming to have a reproducible setup that does not require an interactive/online setup phase. Whether that's possible or not it's one question, but if it was possible it would be a good thing to have.

> > connectivity problems can be debugged separately without affecting the process for everyone else,
> 
> Each operator should check its connectivity before running the share command with the `drand util check <address>` CLI.

This only checks that connection initiated in A gets to B. It does not check if connection initiated in B can get to A.

**Comment by nikkolasg** [comment]
> Ok I missed this part. Is there a way to write a program that: given N public keys ouputs the group.toml file in an offline mode?
> In a way that no one has to run the share command?

Previous version of drand worked like this: I collected all public keys from operators via email or else, then I created the group file manually, then gave it to operators manually, then operators ran the share command by giving the group file.
It led to a huge amount of operational mistakes and confusions. 
Operators still need to run a "reshare" command, it's not done automatically. 

>As long as the operator has not restarted the process, or it has died itself for some reason.

Yes availability is required during the "setup" phase at least for the coordinator - Is this assumption that is bothering you ? I'm not sure how to not require it...

>This is suboptimal because it requires access from the operator to the machines which are supposed to be super safe to run commands.

I see. I feel there is a middle ground to be achieved but it's not clear yet in my mind - how to require the least amount of operations from operators while still allowing automatic deployment, i.e. prevent manual CLI typing.

> Yes, in sync because a node needs to start DKG before anyone else, and that node needs to survive through the process and no node can be correctly configured without having to be online at the same time as the other nodes.

A node, if it wants to participate in the DKG must be online yes. Whether other are online or not doesn't matter from the perspective of one node. 
Are you saying you want the ability of a node to participate in the setup+DKG  phase while not being online all the time ?
